### DISCUSSION TITLE
Is a Theia-created IDE supposed to be keyboard-operable?

### DISCUSSION THREAD

**Discussion by chrisjj** [eclipse-theia/theia#9500] [Q&A]
I ask because I'm having some difficulties. E.g. In Theia Browser Example, I can't find a key to open the File menu. On Help About I can't find a key to scroll the list. On Keyboard Shortcuts, I can't find a key to select the item edit (ENTER edits the first on the list despite that it is not visibly selected).

**Comment by chrisjj** [comment]
PS I'd like to label this Question, but that requires a [githib checkbox](https://help.github.com/en/github/managing-your-work-on-github/applying-labels-to-issues-and-pull-requests) that I cannot find.

**Comment by vince-fugnitto** [comment]
Just an an FYI, you can open the `keyboard shortcuts` view the platforms list of commands and their respective keybindings.  The view can be opened from the file menu + settings, or by using the command palette (<kbd>F1</kbd>) and searching for keyboard shortcuts.

**Comment by chrisjj** [comment]
Thanks, but I don't see any of those key commands under Keyboard Shortcuts - as expected, since they'd normally be standard key commands rather than shortcuts.

By the way, Keyboard Shortcuts is a misnomer for that dialog. Those key commands bindings are rarely shortcuts. Key Bindings would be correct.

**Comment by vince-fugnitto** [comment]
> By the way, Keyboard Shortcuts is a misnomer for that dialog. Those key commands bindings are rarely shortcuts. Key Bindings would be correct.

I believe we simply adopted the naming present in VS Code and did not find enough of a reason to misalign. VS Code also refers to the same view as `Keyboard Shortcuts`.

**Comment by chrisjj** [comment]
Would you say being wrong was sufficient reason?

**Comment by vince-fugnitto** [comment]
> Being wrong is not good enough reason?

Sorry, but I don't see what is wrong about calling a `keyboard shortcuts` view as `keyboard shortcuts`. The view describes all available commands and keybindings, and allows you to update them/revert them to their default values. I think it's more a matter of your own opinion since you also believe a well established editor such as VS Code is wrong as well.

**Comment by chrisjj** [comment]
> Sorry, but I don't see what is wrong about calling a keyboard shortcuts view as keyboard shortcuts. 

Nothing. But it is not a keyboard shortcuts view. Most of the entries aren't keyboard shortcuts. They are simply command-to-key bindings.

> The view describes all available commands and keybindings

Indeed it does.

> I think it's more a matter of your own opinion 

Microsoft's too.

"Keyboard shortcuts are keys or combinations of keys that provide another way to do something that you’d typically do with a mouse." https://support.microsoft.com/en-gb/help/13805/windows-keyboard-shortcuts-in-apps

> since you also believe a well established editor such as VS Code is wrong as well.

I have no doubt the VS Code team is capable of getting something wrong. Every dev team is.

**Comment by paul-marechal** [comment]
AFAIK all of the menus entries are commands. So if you can map keybindings to commands, you effectively can create any "shortcut" you want, since you won't need to use your mouse ever. You can also hit F1 and get an exhaustive list. In practice I don't know if everything is really covered by commands, but I'd say you can do a lot already. Then there's UI navigation like pressing tab to focus inputs and whatnot, we try our best but maybe it doesn't always work. PRs are welcome if anything is missing or bogus.

**Comment by chrisjj** [comment]
> AFAIK all of the menus entries are commands.

But not all commands are menu entries. Or accessible except by bound key.

That's why "keyboard shortcuts" isn't a list of keyboard _short_ cuts. [<< Space is workaround for GitHub formatting fail].

What it is, is a list of commands (including those having no keys https://i.imgur.com/pKMI1Fz.png) allowing them to be bound to keys.

**Comment by paul-marechal** [comment]
Ok, I did not get that you were talking about the title of the widget. Feel free to rename it.

**Comment by chrisjj** [comment]
For avoidance of doubt, it is this:

![image](https://user-images.githubusercontent.com/6157073/73597284-2b8bac80-4522-11ea-867c-df0e448fbe53.png)

**Comment by vince-fugnitto** [comment]
> I have no doubt the VS Code team is capable of getting something wrong. Every dev team is.

While it’s true that any dev team can make mistakes it is still a matter of your own opinion. If you truly believe there should be an update to make to the default naming of the view then feel free to open a new issue describing why it should be updated and get feedback. Else, it is entirely possible to customize the name of the view in your own application as such things are updatable.

**Comment by tsmaeder** [comment]
@chrisjj @vince-fugnitto I got a bit confused by the discussion about terminology, so could you answer me this: can Theia be operated using only the keyboard? Because if not, that would be an accessibility issue.

**Comment by akosyakov** [comment]
Ideally it should be, if not please open individual issues.

**Comment by chrisjj** [comment]
@tsmaeder "could you answer me this: can Theia be operated using only the keyboard?"

No. Since posting I've concluded there's no way to engage menus using keyboard, except for special cases such as the editor context menu.

**Comment by chrisjj** [comment]
> Ideally it should be,

Is there some design document establishing that keyboard operability is intended?

**Comment by paul-marechal** [comment]
@chrisjj I don't think there is such a written requirement about the platform. But this is a collaborative project, each contributor is free to satisfy any of his own requirements as long as it benefits everyone and goes along the overall framework's goals. Keyboard operability seems like a beneficial thing to support, but maybe no-one prioritized this yet. Again: PRs are welcome.

> [...] Since posting I've concluded there's no way to engage menus using keyboard, except for special cases such as the editor context menu.

If you are talking about menus such as "File", "View", etc... Each entry is a command, so you can hit F1 and search for what you would usually click on. I don't see what the problem is.

**Comment by chrisjj** [comment]
> I don't think there is such a written requirement about the platform.

Thanks.

> But this is a collaborative project, each contributor is free to satisfy any of his own requirements as long as it benefits everyone and goes along the overall framework's goals. 

So, if one contributor did add full keyboard operability, the fact the overall framework's goals does not include keyboard operability allows that to be undermined by subsequent contributions of mouse-only features. Right?

> If you are talking about menus such as "File", "View", etc... Each entry is a command, so you can hit F1 and search for what you would usually click on.

Normally on a Windows program, Alt+F shows the File menu with selected commands ordered and grouped by function. On Theia, you get the same from a mouse click. But on keyboard, you're suggesting hit F1 and  type 'file:' to see only some of the File commands, some others too, and not ordered or grouped by function.

> I don't see what the problem is.

It's worth trying Alt-F on a Windows program to see how inferior is your suggested substitute.

**Comment by paul-marechal** [comment]
> So, if one contributor did add full keyboard operability, the fact the overall framework's goals does not include keyboard operability allows that to be undermined by subsequent contributions of mouse-only features. Right?

Like Anton said "ideally" navigation should be as convenient as possible, this would include keyboard shortcuts, maybe. If you are an expert in shortcuts-design and wish to bring Theia's standard up, then you can implement your solution, and also document a way for other contributors to follow in order to properly build on top of it. In order to enforce this even more, one would have to carefully review PRs and call out when keyboard navigatibility could be improved.

If it is missing in some places, we can always try to add it later. So really there are only solutions.

> [...] you're suggesting hit F1 and type 'file:' to see only some of the File commands, some others too, and not ordered or grouped by function.

Yes, this currently works well. Just type what you think the name of command is, good discoverability this way. If you know specific commands in advance its even faster, and you can also bind a shortcut for it.

**Comment by akosyakov** [comment]
> Normally on a Windows program, Alt+F shows the File menu with selected commands ordered and grouped by function. On Theia, you get the same from a mouse click. But on keyboard, you're suggesting hit F1 and type 'file:' to see only some of the File commands, some others too, and not ordered or grouped by function.

Please file an issue with it without the rest of noise.

**Comment by akosyakov** [comment]
https://github.com/eclipse-theia/theia/issues/7063 - opened for it, do you have something else concrete to propose?

@marechal-p I would be fine to add to our review checklist a check to review for proper keyboard accessibility. Could you send a PR? For now proper means that it is operable somehow, we can decide on concrete rules after gaining some experience during reviews and include in the coding guidelines later.

**Comment by chrisjj** [comment]
> > So, if one contributor did add full keyboard operability, the fact the overall framework's goals does not include keyboard operability allows that to be undermined by subsequent contributions of mouse-only features. Right?
 
> Like Anton said "ideally" navigation should be as convenient as possible, this would include keyboard shortcuts, maybe. If you are an expert in shortcuts-design and wish to bring Theia's standard up, then you can implement your solution, and also document a way for other contributors to follow in order to properly build on top of it. In order to enforce this even more, one would have to carefully review PRs and call out when keyboard navigatibility could be improved.

I'll take that as a Yes. Thanks.

I don't have faith that keyboard operability would be maintained by "document a way for other contributors to follow in order to properly build on top of it". It would need a solid policy.

> Yes, this currently works well. Just type what you think the name of command is, good discoverability this way. If you know specific commands in advance its even faster, and you can also bind a shortcut for it.

I don't believe that's any substitute for menus. Hence that so many programs have menus and so few have command palettes - including major Theia IDEs such as Arm Mbed Studio.

**Comment by chrisjj** [comment]
> Please file an issue with it without the rest of noise.

I see you did so. Thanks.

**Comment by chrisjj** [comment]
See https://github.com/eclipse-theia/theia/issues/7082 .

**Comment by vince-fugnitto** [comment]
> I don't believe that's any substitute for menus. Hence that so many programs have menus and so few have command palettes - including major Theia IDEs such as Arm Mbed Studio.

I fail to see the point here. The framework permits downstream applications for benefitting from both **menus** and the **command-palette** (popularized in vscode) in order to perform operations. The **command-palette** is not a replacement for menus but it is another manner in which commands (especially recently used ones) can be easily triggered.

The goal is to make menus more easily accessible when using a keyboard which is a fair point and is why #7082, but I also see the value in using the **command-palette**. 

You brought up Arm's Mbed Studio, do they do something differently with menus?

**Comment by chrisjj** [comment]
> The **command-palette** is not a replacement for menus

I'd taken you to be suggesting otherwise in your answer to my

>> Since posting I've concluded there's no way to engage menus using keyboard, except for special cases such as the editor context menu.

that being

> If you are talking about menus such as "File", "View", etc... Each entry is a command, so you can hit F1 and search for what you would usually click on. I don't see what the problem is.

Thanks for the clarification.

> The goal is to make menus more easily accessible when using a keyboard 

That statement would be useful if they were accessible at all using a keyboard, but since they are not,. I suggest "The goal is to make menus accessible by keyboard.

> I also see the value in using the **command-palette**.

So do I. Just that I don't think that value is of any relevance to menus, except as a temporary workaround.

> You brought up Arm's Mbed Studio, do they do something differently with menus?

I brought up Arm Mbed Studio as an example that does the same. The only difference I see is it exacerbates the issue by disallowing access to the command palette and keyboard bindings editor. 

If there's any better major Theia-made IDE (and I really hope there is), please say.

**Comment by akosyakov** [comment]
> I brought up Arm Mbed Studio ans example that does the same. The only difference I see is it exacerbates the issue by disallowing access to the command palette and keyboard bindings.

cc @thegecko @westbury just in case if you did not follow this discussion

**Comment by thegecko** [comment]
Thanks @akosyakov we are aware of this issue.

**Comment by gtritchie** [comment]
An example of a SPA that does have a traditional main menu accessible via keyboard, look at Google Docs. You can use Ctrl+Alt+F to get to file menu, for example.

**Comment by akosyakov** [comment]
@gtritchie true, i've mentioned it in https://github.com/eclipse-theia/theia/issues/7063. Active committers don't have capacity for it right now, we will have a look when we have. In the meantime anybody is welcomed to send a PR.
